google-cloud-core/lib/google/cloud/config.rb

```diff
@@ -491,9 +486,6 @@ def self.reload_configuration!
           config.add_alias! :project, :project_id
           config.add_field! :credentials, default_creds, match: Object
           config.add_alias! :keyfile, :credentials
-          config.add_field! :service_name, nil, match: String
```

Comment:
Why are `service_name` and `service_version` removed from the top-level config here? Aren't they likely to be used in other gems besides debugger?

Reply:
Yes, those toplevel fields are used in the error reporting gem, and cannot be removed.

Reply:
My intention here is to have both gems check if those fields have been configured, and add if they have not been configured. This would allow users to use only on gem without the other.

Reply:
We can have both gems check-and-add the field, but I don't understand the benefit over having a central registry for toplevel config fields. The debugger and error reporting gems already do not depend on each other. Both depend on google-cloud-core, but this won't change that.

Reply:
I would like to avoid the issue where wanting to add additional fields to the top-level necessitates a new release of the google-cloud-core gem. To me, it seems cleaner to have each gem that uses a top-level configuration responsible for adding it.

Reply:
I'd also consider adding the configuration to the `google-cloud-debugger.rb` file that google-cloud-core uses as a simple plug-in mechanism. That way you could configure the gem before you require it explicitly.

```ruby
require "google/cloud"

Google::Cloud.configure do |config|
  # do things
end

debugger = Google::Cloud.debugger # requires "google/cloud/debugger"
```

Reply:
I see. That sounds reasonable. I take it the intent is to keep `project_id` and `credentials` in core, and move all the others (e.g. `use_logging`) out. It would probably be worth putting a note in the documentation in core that individual gems might add their own toplevel configs.

Reply:
Agreed on the documentation. I will make these changes for the other gems so you can see what that would look like then.

Reply:
Allowing config before requiring the gem might be interesting. I'm not sure how you'd do it, though—and it might be the case in some cases that the gem needs to be loaded in order to compute default values.

Reply:
Give it a consideration. AFAICT the computing of default values (like `default_project_id` and `default_credentials`) are called and happen in the library's constructor, so the files that contain them would already be loaded. But the configuration default values (non-computed variety) are all environment variable lookups right now.

Moving the configuration to those files would complicate use of the the `reload_configuration!` methods that we use in tests. This is partially why I favored using lambdas for the config defaults, since that would allow the lambdas to be called in the tests, where they could mock `ENV`. But I think we can continue to refine to an approach that would allow for more up-front configuration.